Modules/clinic/itertoolsmodule.c.h

```diff
@@ -198,16 +198,16 @@ itertools__grouper(PyTypeObject *type, PyObject *args, PyObject *kwargs)
     PyObject *parent;
     PyObject *tgtkey;
 
-    if ((type == &_grouper_type ||
-         type->tp_init == _grouper_type.tp_init) &&
+    if ((type == clinic_state()->_grouper_type ||
+         type->tp_init == clinic_state()->_grouper_type->tp_init) &&
         !_PyArg_NoKeywords("_grouper", kwargs)) {
         goto exit;
     }
     if (!_PyArg_CheckPositional("_grouper", PyTuple_GET_SIZE(args), 2, 2)) {
         goto exit;
     }
-    if (!PyObject_TypeCheck(PyTuple_GET_ITEM(args, 0), &groupby_type)) {
-        _PyArg_BadArgument("_grouper", "argument 1", (&groupby_type)->tp_name, PyTuple_GET_ITEM(args, 0));
+    if (!PyObject_TypeCheck(PyTuple_GET_ITEM(args, 0), clinic_state()->groupby_type)) {
+        _PyArg_BadArgument("_grouper", "argument 1", (clinic_state()->groupby_type)->tp_name, PyTuple_GET_ITEM(args, 0));
```

Comment:
Here's another performance bottleneck:

Currently, this C function is the only place where the `clinic_state` macro is actually used in the generated clinic code, and it is called four (!) times, potentially calling `PyType_GetModuleByDef` up to four times[^1]. It would be nice if we could add a custom preamble/clean-up to the generated code for a particular function. Maybe we should add such a feature to AC before continuing with this PR.

cc. @rhettinger 

[^1]: the compiler may end up optimalising this

Reply:
Instead of changing Argument Clinic, we can simply these checks out in `itertools__grouper_impl()´ , similar to the implementations in `chain_new()` and `islice_new()`.

Reply:
Thinking about this further, we don't need new clinic functionality; what we need is already in place in AC. What we need is to improve the generated code. I'll open an AC issue and suggest an improvement.

Reply:
Possible improvement to AC is now up as a PR:
-  #101411

Reply:
This will be improved when gh-101409 has landed. Marking this as resolved.